frame/staking/src/pallet/impls.rs

```diff
@@ -629,54 +639,79 @@ impl<T: Config> Pallet<T> {
-	pub fn get_npos_voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		let weight_of = Self::slashable_balance_of_fn();
-		let mut all_voters = Vec::new();
+	pub fn get_npos_voters(
+		maybe_max_len: Option<usize>,
+	) -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
+		let nominator_count = CounterForNominators::<T>::get() as usize;
+		let validator_count = CounterForValidators::<T>::get() as usize;
+		let all_voter_count = validator_count.saturating_add(nominator_count);
+		drop(validator_count);
+		drop(nominator_count);
 
-		let mut validator_count = 0u32;
-		for (validator, _) in <Validators<T>>::iter() {
+		let max_allowed_len = maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count);
+		drop(all_voter_count);
+
+		let mut all_voters = Vec::<_>::with_capacity(max_allowed_len);
+
+		// first, grab all validators, capped by the maximum allowed length.
+		let mut validators_taken = 0u32;
+		for (validator, _) in <Validators<T>>::iter().take(max_allowed_len) {
 			// Append self vote.
-			let self_vote = (validator.clone(), weight_of(&validator), vec![validator.clone()]);
+			let self_vote =
+				(validator.clone(), Self::weight_of(&validator), vec![validator.clone()]);
 			all_voters.push(self_vote);
-			validator_count.saturating_inc();
+			validators_taken.saturating_inc();
 		}
 
-		// Collect all slashing spans into a BTreeMap for further queries.
+		// .. and grab whatever we have left from nominators.
+		let nominators_quota = max_allowed_len.saturating_sub(validators_taken as usize);
 		let slashing_spans = <SlashingSpans<T>>::iter().collect::<BTreeMap<_, _>>();
-
-		let mut nominator_count = 0u32;
-		for (nominator, nominations) in Nominators::<T>::iter() {
-			let Nominations { submitted_in, mut targets, suppressed: _ } = nominations;
-
-			// Filter out nomination targets which were nominated before the most recent
-			// slashing span.
-			targets.retain(|stash| {
-				slashing_spans
-					.get(stash)
-					.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
-			});
-
-			if !targets.is_empty() {
-				let vote_weight = weight_of(&nominator);
-				all_voters.push((nominator, vote_weight, targets));
-				nominator_count.saturating_inc();
+		let mut nominators_taken = 0u32;
+		for nominator in T::SortedListProvider::iter().take(nominators_quota) {
+			if let Some(Nominations { submitted_in, mut targets, suppressed: _ }) =
+				<Nominators<T>>::get(&nominator)
+			{
+				targets.retain(|stash| {
+					slashing_spans
+						.get(stash)
+						.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
+				});
+				if !targets.len().is_zero() {
+					all_voters.push((nominator.clone(), Self::weight_of(&nominator), targets));
+					nominators_taken.saturating_inc();
+				}
+			} else {
+				log!(error, "invalid item in `SortedListProvider`: {:?}", nominator)
 			}
 		}
 
+		// all_voters should have not re-allocated.
+		debug_assert!(all_voters.capacity() == all_voters.len());
```

Comment:
this debug assert is a bit weird, what if there was less voter than the limit ? then the capacity should be less than the less
```suggestion
		debug_assert!(all_voters.capacity() == max_allowed_len);
```

Reply:
0ca98f953341ff5173aaf1d2551b2684d834a8a7

Reply:
no I think we should assert that the capacity hasn't change, so `==max_allowed_len` otherwise the assert is always true and the comment above inacurrate.